src/diffusers/models/adapter.py

```diff
@@ -0,0 +1,246 @@
+                            block_out_channels[i - 1],
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            down=True,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+                elif j == num_res_blocks - 1:
+                    self.body.append(
+                        BottleneckResnetBlock(
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            block_out_channels[i],
+                            down=False,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+                else:
+                    self.body.append(
+                        BottleneckResnetBlock(
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            block_mid_channels[i],
+                            down=False,
+                            ksize=kerenl_size,
+                            proj_ksize=proj_kerenl_size,
+                            sk=res_block_skip,
+                            use_conv=use_conv,
+                        )
+                    )
+        self.body = nn.ModuleList(self.body)
+
+        if block_mid_channels[0] == block_out_channels[0]:
+            # follow standar adapter schema, using fix kernel size 3 for stem conv layer
+            self.conv_in = nn.Conv2d(channels_in * input_scale_factor**2, block_mid_channels[0], 3, 1, 1)
+        else:
+            # if block_mid_channels[i] < block_out_channels[i](bottleneck downsample block), using light weight adapter schema instead
+            self.conv_in = nn.Conv2d(
+                channels_in * input_scale_factor**2,
+                block_mid_channels[0],
+                proj_kerenl_size,
+                1,
+                proj_kerenl_size // 2,
+            )
+
+    def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
+        r"""
+        Args:
+            x (`torch.Tensor`):
+                (batch, channel, height, width) input images for adapter model, `channel` should equal to
+                `channels_in`.
+        """
+        # unshuffle
+        x = self.unshuffle(x)
+        # extract features
+        features = []
+        x = self.conv_in(x)
+        for i in range(self.num_downsample_blocks):
+            for j in range(self.num_res_blocks):
+                idx = i * self.num_res_blocks + j
+                x = self.body[idx](x)
+            features.append(x)
+
+        return features
+
+
+class MultiAdapter(ModelMixin):
```

Comment:
This is probably the best kind of dispatching I have seen in a while. Thank you!